### PR TITLE
Add an experimental JAX scaleDotProductAttention API to use XLA cuDNN fusedAttention feature.

### PR DESCRIPTION
**Pull request by Cjkkkk**
* lower to cuDNN fused attention and flash attention in XLA
* support bmm-(scale)-(bias)-(mask)-softmax-(dropout)-bmm pattern
* support gradient
* support custom_partitioning
* support vmap batching

### CODE REVIEW

File: jax/experimental/fused_attention_stableHLO_test.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                            PartitionSpec('p', 'b', None, None, None),
+                            None, 
+                            None))
+            )
+
+            out, grads = pjitter(input_q, input_k, input_v, bias, mask, grad)
+            print(grads[0][0,0,0,0,:20])
+
+if __name__ == "__main__":
+    main()
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO_test.py:124]: Should we call it with/wo bias/mask/dropout and forward only? Just to test all the combination here.
This will also need sharding tests. You can look at https://github.com/NVIDIA/TransformerEngine/blob/main/tests/jax/test_distributed_fused_attn.py for an example.

Comment (nouiz) [reply jax/experimental/fused_attention_stableHLO_test.py:124]: Are you use this file end up tested if we do:
`pytest jax/experimental`? I don't think so, so it won't be executed by the CI.

File: jax/experimental/fused_attention_stableHLO.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+from functools import partial, reduce
+import operator
+from typing import Any, Optional
+import json
+
+import jax
+import jax.numpy as jnp
+from jax import core, dtypes
+from jax.interpreters import xla
+from jax.interpreters import mlir
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO.py:24]: from jax.interpreters import ir, mlir, xla

File: jax/experimental/fused_attention_stableHLO.py [hunk 3]
```diff
@@ -0,0 +1,561 @@
...
+    # {output, scratch, activation*}
+    has_dropout = dropout_rate > 0
+    has_bias, has_mask = variadic_args
+    operands = [query, key, value]
+    if has_mask:
+        operands.append(mask)
+    if has_bias:
+        operands.append(bias)
+    # get custom call name
+    custom_call_name = get_custom_call_name(has_bias, has_mask, has_dropout, False)
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO.py:246]: Should this return the expected number of input and you can assert this?
Just an extra check in case of future changes.

File: jax/experimental/fused_attention_stableHLO.py [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+    
+    # reshape to original shape
+    grad_query = jnp.reshape(grad_query, (*batch_tuple, q_seq_len, num_heads, head_dim))
+    grad_key = jnp.reshape(grad_key, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grad_value = jnp.reshape(grad_value, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grads = (grad_query, grad_key, grad_value)
+    return grads, out_bdims
+
+# custom partitioning
+def get_padded_spec(arg_info):
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO.py:378]: @mattjj Should this function be added to JAX? We use it at many places.

File: jax/experimental/fused_attention_stableHLO.py [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+    query_spec = get_padded_spec(arg_shapes[0])
+    # (*batch, kv_seq, num_head, head)
+    key_spec = get_padded_spec(arg_shapes[1])
+    # keep out sharding same as query sharding since they have same shape
+    out_sharding = NamedSharding(mesh, PartitionSpec(*query_spec))
+    # activation sharding
+    activation_sharding = NamedSharding(mesh, PartitionSpec(*query_spec[:-3], query_spec[-2], query_spec[-3], key_spec[-3]))
+    return (out_sharding, activation_sharding)
+
+def _dot_product_attention_fwd_partition(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO.py:399]: In this function and the one above, you don't force any sharding. Like if all inputs sharding are supported.
I'm pretty sure the kernel don't support all sharding.
You can inspire yourself from the equivalent in TE:
https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions.py#L1822-L1851

File: jax/experimental/fused_attention_stableHLO_test.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    parser.add_argument("--dropout_rate", dest="dropout_rate", type=float, default=0.)
+    parser.add_argument("--bias", action="store_true")
+    parser.add_argument("--mask", action="store_true")
+    parser.add_argument("--mask_type", dest="mask_type", type=str, default=None)
+    parser.add_argument("--fwd_only", dest='fwd_only', action="store_true")
+    parser.add_argument("--xla_dump_to", dest="xla_dump_to", type=str, default=None)
+    args = parser.parse_args()
+   
+    # Set XLA flags with additional flags when xla_dump_to is provided
+    xla_flags = '--xla_gpu_enable_triton_gemm=false --xla_gpu_simplify_all_fp_conversions --xla_gpu_fused_attention_use_cudnn_rng=true --xla_gpu_enable_cudnn_fmha=true --xla_gpu_enable_xla_runtime_executable=false --xla_gpu_fused_attention_use_cudnn_rng=true'
```
Comment (nouiz) [inline jax/experimental/fused_attention_stableHLO_test.py:63]: Can you remove the first 2 flags?
xla_gpu_fused_attention_use_cudnn_rng is duplicated.
Can you also remove xla_gpu_enable_xla_runtime_executable?
Document why all the non FA flags that you use are set.

Most customers won't set those flags by default while using dot_product_attention(). So we need to document/handle correctly that case.

File: jax/experimental/fused_attention_stableHLO.py (hunk 5 above)
Comment (Cjkkkk) [reply jax/experimental/fused_attention_stableHLO.py:399]: Not exactly sure what you meant here. Could you add an example?

File: jax/experimental/fused_attention_stableHLO_test.py (hunk 6 above)
Comment (Cjkkkk) [reply jax/experimental/fused_attention_stableHLO_test.py:63]: removed.

File: jax/experimental/fused_attention_stableHLO_test.py (hunk 1 above)
Comment (Cjkkkk) [reply jax/experimental/fused_attention_stableHLO_test.py:124]: This is just for an example, I will add more tests in the future.

File: jax/experimental/fused_attention_stableHLO.py (hunk 5 above)
Comment (nouiz) [reply jax/experimental/fused_attention_stableHLO.py:396]: Let say the kernel know how to shard only on the batch dimensions.
Now, maybe the inputs will be sharded on the inner dimensions or the sequence dimensions.
So you need to force the replication on all dimensions except the batch dimensions.
This is done by using None in TE.

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+# Create dot_product_attention_fwd_p for forward operation.
+_dot_product_attention_fwd_p = core.Primitive("dot_product_attention_fwd")
+_dot_product_attention_fwd_p.multiple_results = True
+_dot_product_attention_fwd_p.def_impl(partial(xla.apply_primitive, _dot_product_attention_fwd_p))
+_dot_product_attention_fwd_p.def_abstract_eval(_dot_product_attention_fwd_abstract)
+
+mlir.register_lowering(
+    _dot_product_attention_fwd_p,
+    _dot_product_attention_fwd_cuda_lowering,
+    platform="gpu",
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:553]: Should be platform="cuda"

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+    
+# User interface
+def dot_product_attention(query: Array,
+                          key: Array,
+                          value: Array,
+                          scale: float = 1.0,
+                          bias: Optional[Array] = None,
+                          mask: Optional[Array] = None,
+                          is_causal_mask: bool = False,
+                          seed: int = 42,
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:630]: I'm wondering if there is an easy way to make this compatible with JAX PRNG key. @froystig any ideas?

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 9]
```diff
@@ -0,0 +1,672 @@
...
+    scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    output, activation = _dot_product_attention_fwd_p_wrapper.bind(
+        query, key, value, bias, mask, scale=scale, seed=seed, dropout_rate=dropout_rate,
+        variadic_args=variadic_args, is_flash_attention=is_flash_attention,
+        is_causal_mask=is_causal_mask)
+    res = (query, key, value, bias, mask, activation, output)
+    return output, res
+
+def _dot_product_attention_bwd_rule(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, res, grad_output):
+    # {Q, K, V, bias, mask, activation, fwd_output, dO}
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:186]: Comment here is not necessary

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 10]
```diff
@@ -0,0 +1,672 @@
...
+    output_shape = (batch, q_seq_len, num_heads, head_dim)
+    activation_shape = (batch, num_heads, q_seq_len, kv_seq_len)
+    softmax_stat_shape = (batch, num_heads, q_seq_len)
+    if q_seq_len > 512:
+        # is flash attention
+        return (
+            ShapedArray(output_shape, query_dtype),  # output
+            ShapedArray(softmax_stat_shape, jnp.float32),  # softmax_stat
+        )
+    else:
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:229]: Nit: no need for `else`

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+    # {Q, K, V, mask*, bias*}
+    # {output, scratch, activation*}
+    has_dropout = dropout_rate > 0
+    has_bias, has_mask = variadic_args
+    operands = [query, key, value]
+    if has_mask:
+        operands.append(mask)
+    if has_bias:
+        operands.append(bias)
+    # get custom call name
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:283]: Unnecessary comment

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+    # create custom call here
+    out = custom_call(
+        custom_call_name,
+        result_types=result_types,
+        operands=operands,
+        backend_config=backend_config,
+        operand_layouts=default_layouts(*[ir.RankedTensorType(operand.type).shape for operand in operands]),
+        result_layouts=result_layouts,
+    )
+    # dropout scratch memory
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:309]: Nit: "dropout scratch memory" -> "drop scratch memory" (dropout is a bit confusing in context)

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+    return [hlo.transpose(out.results[0], grad_transpose_perm),
+            hlo.transpose(out.results[1], grad_transpose_perm),
+            hlo.transpose(out.results[2], grad_transpose_perm)]
+
+# batcher
+def _check_valid_batch_dims(bdims):
+    for dim in bdims:
+        assert dim in [0, None], \
+            "Currently only support batch_dim in [0, None], " \
+            f"but got {dim=}"
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:392]: Should this raise NotImplementedError?

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+
+def _dot_product_attention_fwd_batcher(batched_args, batch_dims, *, scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    _check_valid_batch_dims(batch_dims)
+    query, key, value, bias, mask = batched_args
+    query_bdim = batch_dims[0]
+    out_bdims = query_bdim, query_bdim
+
+    *batch_tuple, q_seq_len, num_heads, head_dim = query.shape
+    *_, kv_seq_len, _, _ = key.shape
+    batch = reduce(operator.mul, batch_tuple)
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:402]: Nit: rename batch -> new_batch

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+    spec = None if arg_info.sharding is None else arg_info.sharding.spec
+    ndim = arg_info.ndim
+    if spec is None:
+        return (None,) * ndim
+    assert len(spec) <= ndim
+    return spec + (None,) * (ndim - len(spec))
+    
+# fwd custom partition
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:477]: This function shares a lot of code with the `partition` function. Can the logic that computes shardings be shared via a helper function?

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 16]
```diff
@@ -0,0 +1,672 @@
...
+    # (*batch, q_seq, num_head, head)
+    query_spec = _get_padded_spec(arg_shapes[0])
+    # (*batch, kv_seq, num_head, head)
+    key_spec = _get_padded_spec(arg_shapes[1])
+    # keep out sharding same as query sharding since they have same shape
+    out_sharding = NamedSharding(mesh, PartitionSpec(*query_spec))
+    # activation sharding
+    if query_spec[-3] == key_spec[-3]:
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:485]: Do you need to use negative dimensions here?

Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:485]: Isn't `kv_seq` being sharded here not allowed? You would need to `all_gather` the KVs to properly do flash across them.

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 17]
```diff
@@ -0,0 +1,672 @@
...
+# fwd custom partition
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+    # (*batch, q_seq, num_head, head)
+    query_spec = _get_padded_spec(arg_shapes[0])
+    # (*batch, kv_seq, num_head, head)
+    key_spec = _get_padded_spec(arg_shapes[1])
+    # keep out sharding same as query sharding since they have same shape
+    out_sharding = NamedSharding(mesh, PartitionSpec(*query_spec))
+    # activation sharding
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:484]: Do you need to handle mismatched shardings (like if q's heads are sharded and kv's are replicated)? At the very least, we should `raise NotImplementedError` in these cases.

File: jax/_src/cudnn/fused_attention_stableHLO.py [hunk 18]
```diff
@@ -0,0 +1,672 @@
...
+    infer_sharding_from_operands=_dot_product_attention_bwd_infer_sharding_from_operands,
+    partition=_dot_product_attention_bwd_partition)
+
+mlir.register_lowering(_dot_product_attention_bwd_p_wrapper,
+                        mlir.lower_fun(_dot_product_attention_bwd_lower, multiple_results=True))
+
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_fwd_p)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_fwd_p_wrapper)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_bwd_p)
+dispatch.prim_requires_devices_during_lowering.add(_dot_product_attention_bwd_p_wrapper)
```
Comment (sharadmv) [inline jax/_src/cudnn/fused_attention_stableHLO.py:599]: Why are these needed?

**Review by sharadmv — CHANGES_REQUESTED** [review]
Thank you for making this PR!

File: jax/_src/cudnn/fused_attention_stableHLO_test.py
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stableHLO_test.py:1]: I guess I should also add fused_attention_stablehlo.py into jax/BUILD file?

File: jax/_src/cudnn/fused_attention_stableHLO.py (hunk 16 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stableHLO.py:485]: Not sure how many batch dims will be there, but the sequence dim would be the negative -3 in any case.

File: jax/_src/cudnn/fused_attention_stableHLO.py (hunk 18 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stableHLO.py:599]: @nouiz mentioned there might be XLA crashes if not using wrapper primitives. I didn't fully understand and perhaps @nouiz could explain more on this?

Comment (nouiz) [reply jax/_src/cudnn/fused_attention_stableHLO.py:599]: Right now, there should be a crash in XLA without that. It is @mattjj that gave this WAR. Sadly, we didn't had time to hunt and fix the XLA bug.

**Comment by nouiz** [comment]
Does this PR replace https://github.com/google/jax/pull/17957?
If so, we should close the older one.

File: jax/_src/cudnn/fused_attention_stableHLO.py (hunk 16 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stableHLO.py]: I added checks to only allow sharding on sequence dim and head dim.

**Comment by Cjkkkk** [comment]
@sharadmv Hi, any updates on this?

File: tests/fused_attention_stablehlo_test.py [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+os.environ['XLA_FLAGS'] = '--xla_gpu_enable_cudnn_fmha=true --xla_gpu_fused_attention_use_cudnn_rng=true'
+
+import numpy as np
+import jax
+import jax.numpy as jnp
+from jax.sharding import Mesh
+from jax.sharding import PartitionSpec, NamedSharding
+from jax._src import config
+from jax._src import test_util as jtu
+from jax._src.cudnn.fused_attention_stablehlo import dot_product_attention
```
Comment (nouiz) [inline tests/fused_attention_stablehlo_test.py:28]: Is that how we want people to import it?
I think something like: jax.cudnn.... would be better.
We also need the new operation to show up in the documentation.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 20]
```diff
@@ -0,0 +1,728 @@
...
+import jax.numpy as jnp
+from jax import core, dtypes
+from jax.interpreters import mlir, xla
+from jax.interpreters.mlir import ir
+from jaxlib.hlo_helpers import custom_call
+from jax._src.lib.mlir.dialects import hlo
+from jax._src.core import ShapedArray
+
+from jax.experimental.custom_partitioning import custom_partitioning
+from jax.experimental.pjit import pjit
```
Comment (nouiz) [inline jax/_src/cudnn/fused_attention_stablehlo.py:30]: Remove?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 21]
```diff
@@ -0,0 +1,728 @@
...
+  is_cross_attention = q_seq_len != kv_sqe_len
+  # check if attention pattern is supported by flash attention or fused attention
+  if q_seq_len > 512 and kv_sqe_len > 512 and head_dim in [64, 128]:
+    # check if flash attention is supported
+    is_flash_attention = True
+  elif q_seq_len <= 512 and kv_sqe_len <= 512 and head_dim == 64:
+    # check if regular fused attention is supported
+    is_flash_attention = False
+  else:
+    raise NotImplementedError("Unsupported sequence length and head dim.")
```
Comment (nouiz) [inline jax/_src/cudnn/fused_attention_stablehlo.py:217]: Add the values given and the values supported in the error message?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 22]
```diff
@@ -0,0 +1,728 @@
...
+  key_spec = _get_padded_spec(arg_shapes[1])
+  value_spec = _get_padded_spec(arg_shapes[2])
+  if not query_spec == key_spec == value_spec:
+    raise ValueError("Query, key and value should have same sharding.")
+  seq_spec = query_spec[-3]
+  head_spec = query_spec[-1]
+  if seq_spec != None:
+    raise ValueError("Sharding on sequence dim is not allowed.")
+  if head_spec != None:
+    raise ValueError("Sharding on head dim is not allowed.")
```
Comment (nouiz) [inline jax/_src/cudnn/fused_attention_stablehlo.py:551]: Normally, JAX/XLA silently force None sharding on dimensions it doesn't support.
Why do you raise an error here?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 23]
```diff
@@ -0,0 +1,728 @@
...
+                variadic_args=variadic_args, is_flash_attention=is_flash_attention, is_causal_mask=is_causal_mask)
+  return mesh, impl, out_shardings, arg_shardings
+
+# bwd custom partition
+def _infer_bwd_output_sharding(mesh, arg_shapes):
+  # (*batch, q_seq, num_head, head)
+  query_spec = _get_padded_spec(arg_shapes[0])
+  # (*batch, kv_seq, num_head, head)
+  key_spec = _get_padded_spec(arg_shapes[1])
+  # keep grad query sharding same as query sharding
```
Comment (nouiz) [inline jax/_src/cudnn/fused_attention_stablehlo.py:576]: Here you need to raise an error if it tries to shard on that isn't supported or force None sharding on them, as you did in the fwd function.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 24]
```diff
@@ -0,0 +1,728 @@
...
+  activation_sharding = NamedSharding(mesh, PartitionSpec(*query_spec[:-3], query_spec[-2], query_spec[-3], None))
+  return (out_sharding, activation_sharding)
+
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+  return _infer_fwd_output_sharding(mesh, arg_shapes)
+
+def _dot_product_attention_fwd_partition(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
+  # args sharding
+  arg_shardings = tuple([arg_i.sharding for arg_i in arg_shapes])
```
Comment (nouiz) [inline jax/_src/cudnn/fused_attention_stablehlo.py:564]: If we don't raise a error for not supported sharding, we must force the input sharding to one we support.

**Comment by MoFHeka** [comment]
@Cjkkkk 
But other JAX(Flax or Praxis) attention layers was constructed by Einsum kernels, which couldn't' be lowered to cudnn GEMM and the latest cudnn XLA FMHA kernel. When running attention layers in GPU, it could be only transformed to triton kernel according the XLA dump log...

TE currently only supports a limited number of transformer models (such as MOE is difficult to support) and does not yet support LORA SFT. So it may be necessary to optimize the layer composition of the Jax ecosystem.

It would be nice if someone could also change the code in jax wrapper components like praxis, flax, etc., since they are now written in einsum.
BossHi-20240128172529

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 22 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:551]: I don't know what is the best practice here so I just raise an error for not supported sharding.

File: tests/fused_attention_stablehlo_test.py (hunk 19 above)
Comment (Cjkkkk) [reply tests/fused_attention_stablehlo_test.py:28]: I added a package under jax.cudnn and import from jax._src.
Regarding the docs, where would be best to put the doc?

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 24 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:564]: It is checked in `_infer_fwd_output_sharding`.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 23 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:601]: Added.

File: tests/fused_attention_stablehlo_test.py (hunk 19 above)
Comment (nouiz) [reply tests/fused_attention_stablehlo_test.py]: @yashk2810 Any idea about the doc?

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 22 above)
Comment (nouiz) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: Fine for me.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 24 above)
Comment (nouiz) [reply jax/_src/cudnn/fused_attention_stablehlo.py:584]: SG. Thanks.

**Review by superbobry — APPROVED** [review]

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 25]
```diff
@@ -0,0 +1,753 @@
...
+  # check if cuDNN is installed and if cuDNN version is sufficient 
+  check_cudnn_version(is_flash_attention, is_cross_attention)
+  if mask is not None and is_causal_mask:
+    raise ValueError("can not apply a mask and generate a causal_mask at the same time.")
+  variadic_args = (bias is not None, mask is not None)
+  if bias is None:
+    bias = jnp.zeros(0, dtype=query.dtype)
+  if mask is None:
+    mask = jnp.zeros(0, dtype=query.dtype)
+  # TODO: remove this once scale behavior is fixed
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: Yes, here is a fix in xla: https://github.com/openxla/xla/pull/9317. Should be merged soon.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 26]
```diff
@@ -0,0 +1,753 @@
...
+# _dot_product_attention_fwd must have the same func signature as _dot_product_attention
+_dot_product_attention.defvjp(_dot_product_attention_fwd_rule, _dot_product_attention_bwd_rule)
+    
+# User interface
+def dot_product_attention(query: Array,
+                          key: Array,
+                          value: Array,
+                          bias: Optional[Array] = None,
+                          mask: Optional[Array] = None,
+                          scale: float = 1.0,
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:705]: fixed.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: it is merged now.

**Review by superbobry — COMMENTED** [review]
I found a few issues which were hard to spot in the GitHub UI. Please address them in a separate PR once this one is submitted.
